processing/src/main/java/org/apache/druid/segment/RowCountingCursorDecorator.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment;
+
+import org.apache.druid.query.context.ResponseContext;
+import org.joda.time.DateTime;
+
+public class RowCountingCursorDecorator implements Cursor
+{
+  private Cursor cursor;
+  private ResponseContext responseContext;
+
+  public RowCountingCursorDecorator(Cursor cursor, ResponseContext responseContext)
+  {
+    this.cursor = cursor;
+    this.responseContext = responseContext;
+  }
+
+  @Override
+  public ColumnSelectorFactory getColumnSelectorFactory()
+  {
+    return this.cursor.getColumnSelectorFactory();
+  }
+
+  @Override
+  public DateTime getTime()
+  {
+    return this.cursor.getTime();
+  }
+
+  @Override
+  public void advance()
+  {
+    this.cursor.advance();
+    if (this.responseContext != null) {
```

Comment:
Avoid this null check and make responseContext mandatory in the constructor.  If responseContext is optional, add another constructor without it.